working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,752 @@
+
+When `E` (respectively `E<X1, .. Xk>`) is a protected extension type, and
+`e` is an expression with static type `E` (respectively `E<X1, .. Xk>`), 
+it is a compile-time error to perform a type test (`e is T`) or type cast
+(`e as T`) when the target type `T` is `E` (respectively `E<T1, .. Tk>`).
+
+*The rationale is that an extension type that justifies _any_ constructors
+will need to maintain some invariants, and hence it is not helpful to allow
+implicit introduction of any value of that type with a complete lack of
+enforcement of the invariants.*
+
+*For example:*
+
+```dart
+protected extension type nat on int {
+  factory nat(int value) =>
+      value >= 0 ? value : throw "Attempt to create an invalid nat";
+}
+
+void main() {
+  nat n1 = 42; // Error.
+  var n2 = nat(42); // OK at compile time, and at run time.
+}
+```
+
+*The subtyping relationships are illustrated in the following example:*
+
+```dart
+class IntBox {
+  int i;
+  IntBox(this.i);
+}
+
+protected extension type EvenIntBox on IntBox {
+  factory EvenIntBox(int i) =>
+      i % 2 == 0 ? IntBox(i) : throw "Invalid EvenIntBox";
+  factory EvenIntBox.fromIntBox(IntBox intBox) =>
+      intBox.i % 2 == 0 ? intBox : throw "Invalid EvenIntBox";
+  void next() => this.i += 2;
+}
+
+void main() {
+  var evenIntBox = EvenIntBox(42);
+  evenIntBox.next(); // Methods of `EvenIntBox` maintain the invariant.
+  var intBox = evenIntBox as IntBox; // OK statically and dynamically.
+  intBox.i++; // Invariant of `evenIntBox` violated!
+  evenIntBox = intBox as EvenIntBox; // Error, can't cast to EvenIntBox.
+
+  var evenIntBoxes = [evenIntBox]; // Type `List<EvenIntBox>`.
+  evenIntBoxes[0].next(); // Elements typed as `EvenIntBox`, maintain invariant.
+  List<IntBox> intBoxes = evenIntBoxes; // Compile-time error.
+  intBoxes = evenIntBoxes as dynamic; // Run-time error.
+}
+```
+
+
+### Boxing
+
+This section describes the implicitly induced `box` getter of an explicit
+extension type.
+
+*It may be helpful to equip each explicit extension type with a companion
+class whose instances have a single field holding an instance of the
+on-type. So it's a wrapper with the same interface as the extension type.*
+
+Let _E_ be an explicit extension type. The declaration of _E_ implicitly
+induces a declaration of a class `E.class` with the same type parameters
+and members as _E_, subclass of `Object`, with the same direct
+superinterfaces as `E`, with a final field whose type is the on-type of
+_E_, and with a single argument constructor setting that field to the
+argument. A getter `E.class get box` is implicitly induced in _E_, and it
+returns an object that wraps `this`.
+
+In the case where it would be a compile-time error to declare such a member
+named `box`, the member is not induced.
+
+*The latter rule helps avoiding conflicts in situations where `box` is a
+non-hidden instance member, and it allows developers to write their own
+declaration of `box` if needed.*
```

Comment:
No way to unbox? (If the field had a name, we could get the underlying value. Better yet, allow control over the names of the boxing and unboxing getters.)

The type `E.class` is an interface type, right? It has no supertypes except `Object`.

Reply:
The class `E.class` has a method `E get unbox`. `E.class` has `Object` as its direct superclass, and it implements all the types that `E` implements (so this means that `extension E ... implements C ...` ensures that `E` satisfies the interface of `C`, but `E` is not a subtype of `C`, but `E.class` _is_ a completely full-fledged and working subtype of `C`).